storage/parquet/datasource.go

```diff
@@ -417,39 +435,151 @@ func (rs *RecordStream) RunWorkerInternal(ctx context.Context, tx storage.StateT
+						epochStart := time.Date(1970, 1, 1, 0, 0, 0, 0, time.UTC)
+						return epochStart.Add(time.Hour * 24 * time.Duration(v.(int32)))
+					}
+				case logicalType.TIME != nil:
+					convert = func(v interface{}) interface{} {
+						if logicalType.TIME.Unit.IsSetMILLIS() {
+							return time.Millisecond * time.Duration(v.(int32))
+						} else if logicalType.TIME.Unit.IsSetMICROS() {
+							return time.Microsecond * time.Duration(v.(int64))
+						}
+						return v
+					}
+				case logicalType.TIMESTAMP != nil:
+					convert = func(v interface{}) interface{} {
+						value := v.(int64)
+						if logicalType.TIMESTAMP.IsAdjustedToUTC {
+							var seconds int64
+							var nanoseconds int64
+							if logicalType.TIMESTAMP.Unit.IsSetMILLIS() {
+								seconds = value / 1000
+								nanoseconds = (value % 1000) * 1000000
+								return time.Unix(seconds, nanoseconds)
+							} else if logicalType.TIMESTAMP.Unit.IsSetMICROS() {
+								seconds = value / 1000000
+								nanoseconds = (value % 1000000) * 1000
+								return time.Unix(seconds, nanoseconds)
+							}
+						} else {
+							epochStart := time.Date(1970, 1, 1, 0, 0, 0, 0, time.Local)
+							if logicalType.TIMESTAMP.Unit.IsSetMILLIS() {
+								return epochStart.Add(time.Millisecond * time.Duration(value))
+							} else if logicalType.TIMESTAMP.Unit.IsSetMICROS() {
+								return epochStart.Add(time.Microsecond * time.Duration(value))
+							}
+						}
+						return v
+					}
+				case logicalType.INTEGER != nil:
+					convert = func(v interface{}) interface{} {
+						if logicalType.INTEGER.IsSigned {
+							if logicalType.INTEGER.BitWidth == 64 {
+								return v.(int64)
+							} else if logicalType.INTEGER.BitWidth == 32 {
+								return v.(int32)
+							} else if logicalType.INTEGER.BitWidth == 16 {
+								return int16(v.(int32))
+							} else if logicalType.INTEGER.BitWidth == 8 {
+								return int8(v.(int32))
+							}
+						} else {
+							if logicalType.INTEGER.BitWidth == 64 {
+								return uint64(v.(int64))
+							} else if logicalType.INTEGER.BitWidth == 32 {
+								return uint32(v.(int32))
+							} else if logicalType.INTEGER.BitWidth == 16 {
+								return uint16(v.(int32))
+							} else if logicalType.INTEGER.BitWidth == 8 {
+								return uint8(v.(int32))
+							}
+						}
+						return v
+					}
+				case logicalType.JSON != nil:
+					convert = func(v interface{}) interface{} {
+						var object map[string]interface{}
+						if err := json.Unmarshal(v.([]byte), &object); err != nil {
+							log.Printf("couldn't decode '%s' as json in parquet: %s", v.([]byte), err.Error())
+						} else {
+							return object
+						}
+						return v
+					}
+				case logicalType.BSON != nil:
+				case logicalType.UUID != nil:
+					convert = func(v interface{}) interface{} {
+						data := v.([]byte)
+						return fmt.Sprintf("%.8x-%.4x-%.4x-%.4x-%.12x", data[:4], data[4:6], data[6:8], data[8:10], data[10:16])
```

Comment:
I know this is explained in the linked file, but write a small oneliner like "We convert X to format Y" or whatever, so that a person reading this has some understanding of what's going on, without having to read through some readme